### ISSUE TITLE
Non standard installation paths for private headers

### ISSUE THREAD

**Issue by ghisvail** [clMathLibraries/clRNG#3]
Apart from the following public header files:

```
clRNG.version.h
clRNG.h
mrg31k3p.h
mrg32k3a.h
lfsr113.h
philox432.h
```

which are installed under the standard `/usr/include` prefix,

The remaining private headers are installed under a `/usr/cl/include` prefix which is much less common on Linux.

Is there any rationales about why these files are installed here rather than under something like `/usr/include/private` or `/usr/include/clrng` for instance?

**Comment by mungerd** [comment]
_Currently_ the directory structure for header files is the following:
- `<prefix>/include` for `.h` host-side headers (**public host API**);
- `<prefix>/cl/include` for `.clh` device-side headers (**public device API**);
- `<prefix>/cl/include/private` for private `.c.h` device-side headers, not to be included by user code, but included by the `.clh` device-side headers above.

In short, we wanted to make it clear (for the user and for the host compiler) that the device-side headers are not to be accessed by the host compiler.  Because everything under `/usr/include` is accessible to the host compiler, it is generally understood that files under `/usr/include` can be used by standard C/C++ compilers.  Hence, we chose to place the device-side header files, meant to be used exclusively by the OpenCL C compiler, under a path that can not be found anywhere under `/usr/include`, thus hiding them from standard C/C++ compilers.

As a complement of information, the private `.c.h` headers provide most of the implementation for the device API declared in the `.clh` device-side headers.  They are also used by the library source files, so they are required when building the library, but user host code doesn't need access to them (once the library is compiled and installed).

This said, since we have given the OpenCL C headers a `.clh` extension, they wouldn't clash with the standard C headers with a `.h` extension, so all headers could have been installed under `<prefix>/include/clrng`, for example, with a `private` subdirectory.  But this would disrupt the usual semantics of `/usr/include`.

**Comment by ghisvail** [comment]
Thanks for clarifying @mungerd 

Could you elaborate on:

> all headers could have been installed under <prefix>/include/clrng, for example, with a private subdirectory. But this would disrupt the usual semantics of /usr/include.

**Comment by mungerd** [comment]
> all headers could have been installed under `<prefix>/include/clrng`, for example, with a `private` subdirectory

I meant we could possibly use:
- `<prefix>/include/clrng` for both host-side (`.h`) and device-side (`.clh`) headers;
- `<prefix>/include/clrng/private` for private `.c.h` device-side headers.

> But this would disrupt the usual semantics of `/usr/include`.

It is my understanding that most users expect that files under `/usr/include` are meant to be used by standard C/C++ compilers (i.e., can be included from host code).  This is what I mean by _“usual semantics”_.  However, the device-side headers and private headers (`.clh` and `.c.h`) are meant for the OpenCL C compiler only (cannot be included from host code).

I may be wrong about this, though.  I would go for whatever is most intuitive for users.

**Comment by ghisvail** [comment]
> I meant we could possibly use:
> 
>    <prefix>/include/clrng for both host-side (.h) and device-side (.clh) headers;
>    <prefix>/include/clrng/private for private .c.h device-side headers.

This would be more appropriate from a file organization standpoint (all headers are expected to be found _somewhere_ under `/usr/include`).

As long as you provide an appropriate clRNGConfig.cmake and (optionally) clRNG.pc file for the host-side, you are fine from a user point of view. All the user has to do is call `FindPackage(clRNG)` in `CMake` or `pkg-config <args> clrng` in `Makefile` or directly with `gcc`.

FYI, both `clBLAS` and `clFFT` ship these files and they work absolutely fine.

**Comment by mungerd** [comment]
> all headers are expected to be found _somewhere_ under `/usr/include`

I understand your point of view.  The point I was debating is I would rather say “all headers **for standard C and C++** are expected to be found _somewhere_ under `/usr/include`”, hence perhaps not the OpenCL C headers…
I'd rather have some input from @bragadeesh before we change this.

> FYI, both `clBLAS` and `clFFT` ship these files and they work absolutely fine.

We have such a file (providing `find_package(clRNG)`) somewhere, based on those from `clBLAS` and `clFFT`, but it has not been made public yet.

**Comment by ghisvail** [comment]
> We have such a file (providing find_package(clRNG)) somewhere, based on those from clBLAS and clFFT, but it has not been made public yet.

Awesome, then I will wait for these to be available in a future tagged release before resuming my packaging effort. They are quite essential from a packaging standpoint.

**Comment by mungerd** [comment]
Alright.  I first need to know if we change the directory structure.  This will impact on `find_package`.

**Comment by bragadeesh** [comment]
Thanks @ghisvail for bringing this up.
I don't have a problem with the proposal David made on the headers. I like all headers to go into either /usr/include or /usr/include/clrng.
David - if you have the bandwidth please make the changes. Also, its been long, do we have a FindclRNG.cmake ?

**Comment by ghisvail** [comment]
> I like all headers to go into either /usr/include or /usr/include/clrng.

May I vote for `/usr/include/clrng`, considering the volume of header files to be installed? This would reduce file cluttering under `/usr/include`.

**Comment by bragadeesh** [comment]
Let the public host headers be installed in /usr/include and everything else in /usr/include/clrng and /usr/include/clrng/private as appropriate. This will keep it consistent across clFFT and clBLAS for the public headers.

**Comment by mungerd** [comment]
I'd rather go with @ghisvail's preference.  In addition to reducing cluttering, this would avoid possible collisions with headers with the same names from other libraries.

Just to make sure we're on the same page, this means we would have:

```
/usr/include/clrng/mrg31k3p.h
/usr/include/clrng/mrg31k3p.clh
/usr/include/clrng/private/mrg31k3p.c.h
...
```

If you're okay with that, @bragadeesh, I should be able to do that next week.

**Comment by bragadeesh** [comment]
I hadn't thought about the name collision possibility. Given this, I agree with your proposal, David. Please go ahead. Thanks.

**Comment by mungerd** [comment]
Fixed with #6.

**Comment by ghisvail** [comment]
Awesome. Are you planning to tag a new release soon?
